gadgets/trace_tcpconnect/program.bpf.c

```diff
@@ -372,6 +422,62 @@ int BPF_KRETPROBE(ig_inet_stream_connect_ret, int ret)
 	return exit_tcp_connect(ctx, ret);
 }
 
+SEC("kretprobe/tcp_v4_connect")
```

Comment:
Why not to use `kprobe/tcp_v{4|6}_connect`? It seems to me that the src and dst information is already present in the args:

- `sk` already contains the source IP and port: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L218
- `uaddr` contains the dst IP and port: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L216

Using `kprobe/tcp_v{4|6}_connect` will allow us to still get the [src IP/port](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L350) and [dst port](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L352) when there's an error sending the SYN (which is not covered with this PR).

Reply:
For example the source port is still unknown at the start of the function. It is set in https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L299 There `inet_hash_connect` is called which sets the port. See comment https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/inet_hashtables.c#L1164

Reply:
I don't understand what you are suggesting, maybe I missed something.

This `kretprobe/tcp_v4_connect` was introduced because we don't have the port and ip in `exit_inet_stream_connect` when an error happened

Reply:
I'm suggesting anticipating the collection of the src/dst ip and port from `kretprobe/tcp_v{4|6}_connect` to `kprobe/tcp_v{4|6}_connect` because by using `kretprobe/tcp_v{4|6}_connect`, we'll not provide the src IP/port and dst port when there's an error in `tcp_v{4|6}_connect` (while sending the SYN) because they are cleaned up: [here](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L350) and [here](https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L353), respectively.

The problem with applying my suggestion is that we will not provide the src port if there is an error in `inet_stream_connect` because, as you said, the src port is still unknown in `kprobe/tcp_v{4|6}_connect`. So, let's continue with your proposal, but please document what is still missing.

Reply:
We could maybe trace `inet_hash_connect` instead then: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L299
But I am currently not so sure what this function does right after the hash connect: https://elixir.bootlin.com/linux/v6.11/source/net/ipv4/tcp_ipv4.c#L305